File: app/src/main/java/com/example/bervianto/ecceg_rsa_app/ecc/ECCEGMain.java
package com.example.bervianto.ecceg_rsa_app.ecc;

import com.example.bervianto.ecceg_rsa_app.utils.FileUtils;

import java.math.BigInteger;
import java.util.*;

public class ECCEGMain {
    public static void main(String[] args) throws Exception { // unhandled exception

        // Initiate Classes Component
        BigInteger a = new BigInteger("2"), b = new BigInteger("6"), p = new BigInteger("15485867"),
            k = new BigInteger("30");
        // y = x^3 + 2x + 6 mod 15485867
        // this is programmer defined, but needs to be consistent for encrypt and decrypt
        ECC ecc = new ECC();
        ecc.a = a;
        ecc.b = b;
        ecc.p = p;
        ecc.k = k;
        
        // Initiate ended

        System.out.println("Choose an action : ");
        System.out.println("1. Generate Key");
        System.out.println("2. Encrypt File");
        System.out.println("3. Decrypt File");
        System.out.print("Input 1/2/3 = ");
        Scanner s = new Scanner(System.in);
        int chosen = s.nextInt();

        if (chosen == 1) {
            String private_key_filepath = "key.pri";
            String public_key_filepath = "key.pub";
            ECCEGMain.generateKey(ecc, private_key_filepath, public_key_filepath);
        }
        else if (chosen == 2) {
            String public_key_filepath = "key.pub";
            String plainfile = "datatest.txt";
            String ciphfile = plainfile + ".ciph";
            ECCEGMain.encryptFile(ecc, public_key_filepath, plainfile, ciphfile);
        }
        else if (chosen == 3) {
            String private_key_filepath = "key.pri";
            String public_key_filepath = "key.pub";
            String plainfile = "datatest.txt";
            String ciphfile = "datatest.txt.ciph";
            ECCEGMain.decryptFile(ecc, private_key_filepath, ciphfile, plainfile);
        }
        else System.out.println("Wrong choice!!");
    }

    public static void generateKey(ECC ecc, String private_key_filepath, String public_key_filepath) throws Exception {
        ECCEG ecceg = new ECCEG(ecc, ecc.getBasePoint());
        System.out.print("Private key = ");
        System.out.println(ecceg.getPrivateKey());
        ecceg.savePrivateKey(private_key_filepath);
        System.out.println("Saved in " + private_key_filepath);
        System.out.print("Public key = ");
        System.out.println(ecceg.getPublicKey());
        ecceg.savePublicKey(public_key_filepath);
        System.out.println("Saved in " + public_key_filepath);
    }

    public static void encryptFile(ECC ecc, String public_key_filepath, String filepath_plain, 
        String filepath_ciph) throws Exception {
            ECCEG ecceg = new ECCEG(ecc, ecc.getBasePoint());
            ecceg.loadPublicKey(public_key_filepath);
            System.out.println("Public key loaded...");
            byte[] read = FileUtils.getBytes(filepath_plain);
            System.out.println("---===Plainteks===---");
            System.out.println(new String(read));
            System.out.println("---======END======---");
            System.out.println();
            List<Pair<Point,Point>> enc = ecceg.encryptBytes(read);
            System.out.println("---===Cipherteks===---");
            for (Pair<Point,Point> pp: enc) {
                System.out.print(String.format("%02x%02x%02x%02x",
                    pp.left.x.intValue(),
                    pp.left.y.intValue(),
                    pp.right.x.intValue(),
                    pp.right.y.intValue()));
            }
            System.out.println();
            System.out.println("---======END======---");
            FileUtils.savePointsToFile(filepath_ciph, enc);
    }

    public static void decryptFile(ECC ecc, String private_key_filepath,
                                   String filepath_ciph, String destination) throws Exception {
            ECCEG ecceg = new ECCEG(ecc, ecc.getBasePoint());
            ecceg.loadPrivateKey(private_key_filepath);
            System.out.println("Private key loaded...");
            List<Pair<Point,Point>> read_enc = FileUtils.loadPointsFromFile(filepath_ciph);
            List<Point> read_dec = ecceg.decrypt(read_enc);
            System.out.println("---===Plainteks===---");
            for (Point pp: read_dec)
                System.out.print((char)ecc.pointToInt(pp).byteValue());
            System.out.println();
            System.out.println("---======END======---");
    }
}

File: app/src/main/java/id/my/berviantoleo/ecceg_rsa_app/lib/ecc/ECCEGMain.kt
package id.my.berviantoleo.ecceg_rsa_app.lib.ecc

import id.my.berviantoleo.ecceg_rsa_app.utils.FileUtils.getBytes
import id.my.berviantoleo.ecceg_rsa_app.utils.FileUtils.loadPointsFromFile
import id.my.berviantoleo.ecceg_rsa_app.utils.FileUtils.savePointsToFile
import java.math.BigInteger
import java.util.Scanner

object ECCEGMain {
    @Throws(Exception::class)
    @JvmStatic
    fun main(args: Array<String>) { // unhandled exception

        // Initiate Classes Component

        val a = BigInteger("2")
        val b = BigInteger("6")
        val p = BigInteger("15485867")
        val k = BigInteger("30")
        // y = x^3 + 2x + 6 mod 15485867
        // this is programmer defined, but needs to be consistent for encrypt and decrypt
        val ecc = ECC()
        ecc.a = a
        ecc.b = b
        ecc.p = p
        ecc.k = k

        // Initiate ended
        println("Choose an action : ")
        println("1. Generate Key")
        println("2. Encrypt File")
        println("3. Decrypt File")
        print("Input 1/2/3 = ")
        val s = Scanner(System.`in`)
        val chosen = s.nextInt()

        when (chosen) {
            1 -> {
                val privateKeyFilepath = "key.pri"
                val publicKeyFilepath = "key.pub"
                generateKey(ecc, privateKeyFilepath, publicKeyFilepath)
            }
            2 -> {
                val publicKeyFilepath = "key.pub"
                val plainfile = "datatest.txt"
                val ciphfile = "$plainfile.ciph"
                encryptFile(ecc, publicKeyFilepath, plainfile, ciphfile)
            }
            3 -> {
                val privateKeyFilepath = "key.pri"
                val plainfile = "datatest.txt" // This seems to be the intended destination for the decrypted file
                val ciphfile = "datatest.txt.ciph"
                decryptFile(ecc, privateKeyFilepath, ciphfile, plainfile)
            }
            else -> println("Wrong choice!!")
        }
    }

    @JvmStatic
    @Throws(Exception::class)
    fun generateKey(ecc: ECC, privateKeyFilepath: String, publicKeyFilepath: String) {
        val ecceg = ECCEG(ecc, ecc.basePoint!!)
        print("Private key = ")
        println(ecceg.getPrivateKey())
        ecceg.savePrivateKey(privateKeyFilepath)
        println("Saved in $privateKeyFilepath")
        print("Public key = ")
        println(ecceg.publicKey)
        ecceg.savePublicKey(publicKeyFilepath)
        println("Saved in $publicKeyFilepath")
    }

    @Throws(Exception::class)
    private fun encryptFile(
        ecc: ECC, publicKeyFilepath: String, filepathPlain: String,
        filepathCiph: String?
    ) {
        val ecceg = ECCEG(ecc, ecc.basePoint!!)
        ecceg.loadPublicKey(publicKeyFilepath)
        println("Public key loaded...")
        val read = getBytes(filepathPlain)
        println("---===Plainteks===---")
        println(String(read!!))
        println("---======END======---")
        println()
        // Changed to MutableList<Pair<Point?, Point?>> based on error messages
        val enc: MutableList<Pair<Point?, Point?>> = ecceg.encryptBytes(read)
        println("---===Cipherteks===---")
        for (ppPair in enc) { // ppPair is Pair<Point?, Point?> (non-null Pair)
            // Assuming if a pair exists, its components (left and right points) should exist.
            // And if points exist, their coordinates x, y should also exist.
            // If this assumption is incorrect, NullPointerExceptions can occur here.
            // Proper error handling or a redesign of ECCEG.encryptBytes might be needed.
            print(
                String.format(
                    "%02x%02x%02x%02x",
                    ppPair.left!!.x.toInt(),
                    ppPair.left!!.y.toInt(),
                    ppPair.right!!.x.toInt(),
                    ppPair.right!!.y.toInt()
                )
            )
        }
        println()
        println("---======END======---")
        // savePointsToFile expects MutableList<Pair<Point?, Point?>>
        savePointsToFile(filepathCiph, enc)
    }

    @Throws(Exception::class)
    private fun decryptFile(
        ecc: ECC, privateKeyFilepath: String,
        filepathCiph: String, destinationPath: String? // Renamed for clarity, original 'destination'
    ) {
        val ecceg = ECCEG(ecc, ecc.basePoint!!)
        ecceg.loadPrivateKey(privateKeyFilepath)
        println("Private key loaded...")
        // Changed to MutableList<Pair<Point?, Point?>> based on error messages
        val readEnc: MutableList<Pair<Point?, Point?>> = loadPointsFromFile(filepathCiph)
        // Align with actual signature of ecceg.decrypt
        val readDec: MutableList<Point?> = ecceg.decrypt(readEnc)
        println("---===Plainteks===---")
        // TODO: The decrypted content is printed to console. If it should be saved to 'destinationPath', implement file writing here.
        for (pp in readDec) {
            // Assuming if a point exists in the decrypted list, it should be a non-null Point.
            // If pp can be null, then pp!! will throw NPE.
            // If ecc.pointToInt can't handle a Point that might be intrinsically "null" (e.g. point at infinity), that's another issue.
            pp?.let {
                 print(Char(ecc.pointToInt(it).toByte().toUShort()))
            }
        }
        println()
        println("---======END======---")
    }
}